trello-history/trello-history.md

```diff
@@ -0,0 +1,212 @@
+We first approached the problem with the idea to use a
+[web extension](https://developer.mozilla.org/fr/docs/Mozilla/Add-ons/WebExtensions).
+They offer a convenient way to enhance the navigation experience and could be
+written using traditionnal browser side web technologies (HTML, CSS, JS). So we
+created a POC during our last winter hackathon in January, to demonstrate what
+we intended to do was possible:
+- Access a card history,
+- render it on the back of that card _with markdown support_,
+- randomly insert unicorns in the background.
+
+No black magic behind the scene. We simply used the Trello native
+[client](https://developers.trello.com/docs/clientjs) to access their API with a
+good old XmlHttpRequest. To be called, the API requires explicit authentication:
+1. An auto generated API key,
+2. a token that is generated once the end user granted the API access, on behalf
+of the user identified by this API key.
+
+Although it quite worked well, it also had several drawbacks. Because Trello is
+an SPA and we decided to use a web extension, we had trouble dealing with
+asynchronicity. Indeed, the Trello API only let us perform actions on domain
+oriented objects (card, boards, members...). That is, we still had no way to
+directly interact with the Trello front application, nor to listen for events
+such as card loading success to know **when** we should get and display it's
+history.
+
+We considered to get arround this problem by observing
+[DOM Mutations](https://developer.mozilla.org/fr/docs/Web/API/MutationObserver),
+and detect when a specific element entered the DOM, identified by it's class or
+id. Oh listen, what's that sound ? It looks like the WTF alarms are ranging
+afar ! Relying on HTML attributes to identify elements, which are very likely
+to change over time, is a 100% chance for this project to end in the cemetery of
+repositories nobody dares to debug. We felt the time for a perspective change
+has come.
+
+## An implementation change
+
+Trello offers another possibility to build features upon their application, with
+what they so called [Power ups](https://trello.com/power-ups). See them as
+external libraries which communicates with third services like Google, Facebook
+or the one you built with your little fingers. An important limitation though,
+these power ups are tied to specific locations opened by Trello, such as
+`card-back`, `card-buttons`... These locations are designated by _capabilities_
+in the their vocabulary.
+
+![](https://github.com/jaljo/articles/raw/710f8e2d3561a5fa65bc6422b99aad88695871f3/trello-history/images/pu-setup.png)
+
+**Capabilities are set in the Trello Power-ups administration board. Notice only
+administrators can create and expose power ups to their teams.**
+
+On bottom of that form, the _iframe connector url_ targets where the application
+of the power up is hosted. As you can see, it must be served over HTTPS, which
+can be a bit tricky when working in development environment.
```

Comment:
You can speak about the Power ups button on trello and the feature to delete data (convenient when testing authorization ;) )